src/main/java/org/wildfly/security/auth/spi/Attributes.java

```diff
@@ -576,13 +598,68 @@ default boolean isEmpty() {
     }
 
     /**
-     * Get the collection of values for the given key.  The result may implement {@link SetEntry} if the values
-     * are distinct (for example, a role or group set).
+     * Returns a read-only instance of this instance.
      *
-     * @param key the attribute name
-     * @return the (possibly empty) attribute value collection
+     * @return a read-only instance of this instance.
      */
-    Entry get(String key);
+    default Attributes asReadOnly() {
+        return new Attributes() {
+            private Map<String, Entry> entryCache = new HashMap<>();
+            private Collection<Entry> cachedEntries;
+            Set<String> keySet = Attributes.this.keySet();
```

Comment:
What I mean is, this could be inlined completely, so instead of accessing the keySet field, you just call Attributes.this.keySet() every time.